Update wording in Geospatial spec (#560)
apache/parquet-format: Geospatial.md

The default CRS `OGC:CRS84` means that the geospatial features must be stored
in the order of longitude/latitude based on the WGS84 datum.

> Non-defaullt CRS values are specified by any string that uniquely identifies coordinate reference system associated with this type.

Reviewer: ```suggestion
Non-default CRS values are specified by any string that uniquely identifies coordinate reference system associated with this type.
```

> Non-defaullt CRS values are specified by any string that uniquely identifies coordinate reference system associated with this type.

Reviewer: ```suggestion
Non-defaullt CRS values are specified by any string that uniquely identifies a coordinate reference system associated with this type.
```

> To maximize interoperability suggested (but not limited to) formats for CRS are:

Reviewer: ```suggestion
To maximize interoperability, the following Coordinate Reference System (CRS) formats are recommended (non-exhaustive):
```

> * `<projjson>` - A complete CRS definition embedded directly using the [PROJJSON](https://proj.org/en/stable/specifications/projjson.html) specification. Example for `OGC:CRS83`: `{"$schema": "https://proj.org/schemas/v0.7/projjson.schema.json","type": "GeographicCRS","name": "NAD83 (CRS83)","datum": {"type": "GeodeticReferenceFrame"...`

Reviewer 2: so readers first have to determine if the value is a valid projjson? And if it isn't that try to parse the key-value pair (or vice versa)?  Maybe we should be explicit about this guidance?

Author: Given that these are only a suggested formats - not sure how to word this without sounding restrictive.

Readers would have to choose what formats they support - whether its superset of suggested or subset of suggested formats.

Algorithm for suggested formats would be:
1) If string starts with `projjson:` - go with `projjson:<key_name>`
2) If string starts with `srid:` - go with `srid:<identifier>`
3) Try parse string as a projjson - if it fails
4) Load string as `authority:code`

Author: Do you think we should add this guidance ?

Reviewer 3: I would also leave it to the readers to decide if they want to do a full parse on the PROJJSON string. Regarding how to parse the field overall, I think what @milastdbx is basically the logic. I do not see any ambiguity. The PROJJSON string should start with `{` anyways which is different from any other option listed.

Reviewer 2: Agree there is no ambiguity, as someone not intimately familiar with the the Geo types, it might be a useful summary, if the general consensus is this isn't, then I don't feel too strongly on adding it.

Reviewer 4: Since this should be quite obvious how to implement, my feeling is that we do not need to specify exactly how a reader is supposed to parse those fields (like, there would be nothing wrong with a reader that first checks if the string is projjson, before trying to check for `srid:..`, instead of the other way around)

Author: If I read everyones comments correctly, general conensus seems that its clear how readers should parse it without being explicit in the spec.

If this is the case - @emkornfield can we proceed with the merge ?

Reviewer 2: Yes, I'll send a final note to the mailing list and aim to merge tomorrow if there are no objections.

* `inlined_projjson` - Inlining whole CRS definition in [PROJJSON](https://proj.org/en/stable/specifications/projjson.html) format.
* `srid:identifier` - [SRID - Spatial reference identifier](https://en.wikipedia.org/wiki/Spatial_reference_system#Identifier), `identifier` is the SRID itself.

> * `projjson:table_property_name` - where `table_property_name` is the name of a table property where the projjson string is stored

Reviewer: ```suggestion
* `authorithy:identifier`- A reference using a well-known authority and its identifier. For example: `OGC:CRS83`.
* `srid:identifier` - A reference using a Spatial Reference Identifier (SRID), where identifier is the numeric SRID value. For example: `SRID:0`.
* `projjson` - A complete CRS definition embedded directly using the [PROJJSON](https://proj.org/en/stable/specifications/projjson.html) specification.
* `projjson:table_property_name` - A reference to a CRS definition stored as a table property, where `table_property_name` identifies the property containing the [PROJJSON](https://proj.org/en/stable/specifications/projjson.html) string.
```

> * `projjson:<table_property_name>` - where `<table_property_name>` is a reference to a CRS definition in [PROJJSON](https://proj.org/en/stable/specifications/projjson.html), stored in a file level key/value metadata key or table property.

Reviewer 4: ```suggestion
* `projjson:<table_property_name>` - where `<table_property_name>` is a reference to a CRS definition in [PROJJSON](https://proj.org/en/stable/specifications/projjson.html), stored in a file level key/value metadata key.
```

For Parquet, it is just the file-level metadata (I don't think the spec talks about "table properties")

Author: I reworded it, LMK what you think

* `projjson:table_property_name` - where `table_property_name` is a reference to a CRS definition in [PROJJSON](https://proj.org/en/stable/specifications/projjson.html), stored in the table property.
* `projjson` - A complete CRS definition embedded directly using the [PROJJSON](https://proj.org/en/stable/specifications/projjson.html) specification. Example for OGC:CRS83:
```
{
  "type": "GeographicCRS",
  "name": "NAD83",
  "datum": {
    "type": "GeodeticReferenceFrame",
    "name": "North American Datum 1983",
    "ellipsoid": {
      "name": "GRS 1980",
      "semi_major_axis": 6378137,
      "semi_minor_axis": 6356752.314140356,
      "unit": "metre"
    }
  },
  "coordinate_system": {
    "subtype": "ellipsoidal",
    "axis": [
      {
        "name": "Geodetic longitude",
        "abbreviation": "Lon",
        "direction": "east",
        "unit": "degree"
      },
      {
        "name": "Geodetic latitude",
        "abbreviation": "Lat",
        "direction": "north",
        "unit": "degree"
      }
    ]
  },
  "id": {
    "authority": "EPSG",
    "code": 4269
  }

> }

Author: ```suggestion
{
  "$schema": "https://proj.org/schemas/v0.7/projjson.schema.json",
  "type": "GeographicCRS",
  "name": "NAD83 (CRS83)",
  "datum": {
    "type": "GeodeticReferenceFrame",
    "name": "North American Datum 1983",
    "ellipsoid": {
      "name": "GRS 1980",
      "semi_major_axis": 6378137,
      "inverse_flattening": 298.257222101
    }
  },
  "coordinate_system": {
    "subtype": "ellipsoidal",
    "axis": [
      {
        "name": "Geodetic longitude",
        "abbreviation": "Lon",
        "direction": "east",
        "unit": "degree"
      },
      {
        "name": "Geodetic latitude",
        "abbreviation": "Lat",
        "direction": "north",
        "unit": "degree"
      }
    ]
  },
  "scope": "Not known.",
  "area": "North America - onshore and offshore: Canada - Alberta; British Columbia; Manitoba; New Brunswick; Newfoundland and Labrador; Northwest Territories; Nova Scotia; Nunavut; Ontario; Prince Edward Island; Quebec; Saskatchewan; Yukon. Puerto Rico. United States (USA) - Alabama; Alaska; Arizona; Arkansas; California; Colorado; Connecticut; Delaware; Florida; Georgia; Hawaii; Idaho; Illinois; Indiana; Iowa; Kansas; Kentucky; Louisiana; Maine; Maryland; Massachusetts; Michigan; Minnesota; Mississippi; Missouri; Montana; Nebraska; Nevada; New Hampshire; New Jersey; New Mexico; New York; North Carolina; North Dakota; Ohio; Oklahoma; Oregon; Pennsylvania; Rhode Island; South Carolina; South Dakota; Tennessee; Texas; Utah; Vermont; Virginia; Washington; West Virginia; Wisconsin; Wyoming. US Virgin Islands. British Virgin Islands.",
  "bbox": {
    "south_latitude": 14.92,
    "west_longitude": 167.65,
    "north_latitude": 86.45,
    "east_longitude": -40.73
  },
  "id": {
    "authority": "OGC",
    "code": "CRS83"
  }
}
```

The default CRS `OGC:CRS84` means that the geospatial features must be stored
in the order of longitude/latitude based on the WGS84 datum.

> Non-default CRS values are specified by any case insensitive string that uniquely identifies a coordinate reference system associated with this type.

Reviewer 4: Is there a reason to include the "case insensitive" part here? 
AFAIK the projjson keys are certainly case sensitive

Or maybe only mention this for the relevant options, as the "authority" in `authority:code` can indeed be case insensitive?

Author: thats a good point - ill move it to authority:code section

Author: actually ill remove case-insensitivity altogether, it does not hurt writers thinking that this is case insensitive, its up to readers to be more flexible, and they will do sensible thing

Reviewer 3: Regarding case-insensitivity, I agree it does not make sense for PROJJSON. However, I think it would be helpful to have it for the other three cases. It specifically removes the ambiguity as to whether `SRID:<identifier>` or `srid:<identifier>` are both valid or not, and same for the `<authority>:<code>` and `projjson:<table_property_name>`.

To maximize interoperability, suggested (but not limited to) formats for CRS are:

> * `authority:code` - where `authority` represents some well known authorities, and `code` is the code used by the authority to identify the CRS. Examples are - `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`, `EPSG:3857`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions of coordinate reference systems provided by some well known authorities.

Reviewer 5: Maybe worth adding an ["IGN" authority reference](https://spatialreference.org/ref/ignf/) example, since they use non-numeric `auth_srid` values, which is kind of a unique quirk they have that bungs up code expecting integers only.

Reviewer 3: +1

Although the OGC ones have a non-numeric/alphenumeric code already.

Reviewer 5: True, I guess I'm just still grousing that my 25 year old assumption did not hold.
```
                    Table "public.spatial_ref_sys"
  Column   |          Type           | Collation | Nullable | Default 
-----------+-------------------------+-----------+----------+---------
 srid      | integer                 |           | not null | 
 auth_name | character varying(256)  |           |          | 
 auth_srid | integer                 |           |          | 
 srtext    | character varying(2048) |           |          | 
 proj4text | character varying(2048) |           |          | 
```

Reviewer 3: When we were discussing this part with @milastdbx I specifically suggested to use `<authority>:<code>` vs `<authority>:<identifier>` so that we can allow non-numeric values for the second part.

Reviewer 3: By the way, this was inspired by how PROJJSON refers to the `id` part of the CRS.

Author: added example

> * `authority:code` - where `authority` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions coordinate reference systems provided by some well known authorities.

Author: ```suggestion
* `authority:code` - where `authority` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions of coordinate reference systems provided by some well known authorities.
```

> * `authorithy:code` - where `authorithy` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions coordinate reference systems provided by some well known authorities.

Reviewer: ```suggestion
* `authority:code` - where `authority` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions coordinate reference systems provided by some well known authorities.
```

> * `authorithy:code` - where `authorithy` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions coordinate reference systems provided by some well known authorities.

Reviewer 6: definitions => definitions of

> * `authority:code` - where `authority` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions of coordinate reference systems provided by some well known authorities.

Reviewer 3: Let's add one more EPSG example, for example `EPSG:3857`.

> * `authority:code` - where `authority` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions of coordinate reference systems provided by some well known authorities.

Author: ```suggestion
* `authority:code` - where `authority` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`,  `EPSG:3857`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions of coordinate reference systems provided by some well known authorities.

> * `authority:code` - where `authority` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`,  `EPSG:3857`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions of coordinate reference systems provided by some well known authorities.

Author: ```suggestion
* `authority:code` - where `authority` represents some well known authorities, and `code` is the code used by the authority to identify the CRS  - examples are `OGC:CRS84`, `OGC:CRS83`, `OGC:CRS27`, `EPSG:4326`, `EPSG:3857`. See [https://spatialreference.org/](https://spatialreference.org/) for definitions of coordinate reference systems provided by some well known authorities.

> * `srid:identifier` -  A reference using a [Spatial reference identifier (SRID)](https://en.wikipedia.org/wiki/Spatial_reference_system#Identifier), where identifier is the numeric SRID value. For example: `SRID:0`. 

Reviewer 5: Without an authority, what does an SRID mean? Does the existence of an `SRID:X` reference imply the existence of a `spatial_ref_sys` table to dereference it into something actionable?

Reviewer 3: That's a very good question. From my perspective `SRID:X` means that it is not standard so the engine needs to figure out how to deal with it. Do we have to specify the requirements on the engine side here? My thinking/preference is that we should not do that.

Practically speaking the main reason for this entry is `SRID:0`.

* `projjson:table_property_name` - where `table_property_name` is a reference to a CRS definition in [PROJJSON](https://proj.org/en/stable/specifications/projjson.html), stored in a file level key/value metadata key or table property.

> * `projjson` - A complete CRS definition embedded directly using the [PROJJSON](https://proj.org/en/stable/specifications/projjson.html) specification. Example for OGC:CRS83:

Reviewer 3: ```suggestion
* `projjson` - A complete CRS definition embedded directly using the [PROJJSON](https://proj.org/en/stable/specifications/projjson.html) specification. Example for `OGC:CRS83`:
```

> * `projjson` - A complete CRS definition embedded directly using the [PROJJSON](https://proj.org/en/stable/specifications/projjson.html) specification. Example for `OGC:CRS83`:

Reviewer 4: At the risk of igniting a longer discussion, but since in practice most tools/libraries that have already implemented those native geo types are defaulting to the inline PROJJSON, we might want to put this first in the list?

Author: im fine with this

Author: actually i think this makes spec less readable given long example - not sure wdyt ?

Reviewer 4: I would personally be fine with removing that long example, because even at the end of the list, it breaks the flow of the full document because of the vertical screen estate it takes. IMO the link to the projjson specification is clear enough about what it is. 
And potentially could include a abbreviated example, like `{"$schema": "https://proj.org/schemas/v0.7/projjson.schema.json", "type": "GeographicCRS", "name": "NAD83 (CRS83)", "datum": { ....`

Reviewer 3: +1 on the abbreviated example.

Reviewer 3: We could include a link together with the abbreviated example. Something like:
```
See [here](https://spatialreference.org/ref/ogc/CRS83/projjson.json) for the full PROJJSON definition for this example.
```

Author: Updated. Also i've put it on top of the list

> * `projjson` - A complete CRS definition embedded directly using the [PROJJSON](https://proj.org/en/stable/specifications/projjson.html) specification. Example for `OGC:CRS83`:

Reviewer 4: It might be clear from the fact that the custom CRS value should be a string, but the example below showing the JSON _object_ might make it confusing again. So wondering if there is some formulation to make it clearer this is the string representation of the JSON to be used here.

(I know I had made that mistake in the past for the GeoParquet spec, although there this crs value is included inside a json metadata key:value store as well, so there was this potential ambiguity. That might not actually exist here, since it is a string keyword of the type)

Author: Just before going into suggested formats we say that `CRS` is any **string** value,  so hopefully that makes it unambigous. WDYT ?

What I think might be better is to do somethign like * `<projjson>`  instead of `projjson` making it clear that this is not a literal, but a placeholder for what is about to be described.

However i do not see anybody else in spec doing it this way

Reviewer 4: > What I think might be better is to do somethign like * <projjson> instead of projjson making it clear that this is not a literal, but a placeholder for what is about to be described.

That might actually also make it clearer if used for the other entries as well. For example for `projjson:table_property_name`, to make it clearer that "projjson" is literal and "table_property_name" is a placeholder. So I would be in favor of using it for all entries then

Reviewer 3: So basically?

* `<projjson>`
* `<authority>:<code>`
* `srid:<identifier>`
* `projjson:<table_property_name>`

I also think this is cleaner as to what is expected verbatim and what is a placeholder.

The Z values introduce the third dimension coordinate. Usually they are used to
indicate the height, or elevation.

M values are an opportunity for a geospatial instance to track a value in a

> fourth dimension. These values can be used as a linear reference value (e.g.,

Author: ```suggestion
M values are an opportunity for a geospatial instance to track a value in a 
fourth dimension. These values can be used as a linear reference value (e.g., 
```

highway milepost value), a timestamp, or some other value as defined by the CRS.

Bounding box is defined as the thrift struct below in the representation of
min/max value pair of coordinates from each axis. Note that X and Y Values are
always present. Z and M are omitted for 2D geospatial instances.

When calculating a bounding box, null or NaN values in a coordinate
dimension are skipped. For example, `POINT (1 NaN)` contributes a value to X
but no values to Y, Z, or M dimension of the bounding box. If a dimension has
only null or NaN values, that dimension is omitted from the bounding box. If

> either the X or Y dimension is missing, then the bounding box itself is not

Author: ```suggestion
When calculating a bounding box, null or NaN values in a coordinate 
dimension are skipped. For example, `POINT (1 NaN)` contributes a value to X 
but no values to Y, Z, or M dimension of the bounding box. If a dimension has 
only null or NaN values, that dimension is omitted from the bounding box. If 
either the X or Y dimension is missing, then the bounding box itself is not 
```
